src/ucp/core/ucp_ep.c

```diff
@@ -2408,24 +2435,54 @@ int ucp_ep_config_test_rndv_support(const ucp_ep_config_t *config)
 }
 
 static UCS_F_ALWAYS_INLINE int
-ucp_ep_is_am_keepalive(ucp_ep_h ep, ucp_lane_index_t lane)
+ucp_ep_is_am_keepalive(ucp_ep_h ucp_ep, ucp_rsc_index_t rsc_idx)
 {
     /* if we have ep2iface transport we need to send an active-message based
      * keepalive message to check the remote endpoint still exists */
-    return (ep->flags & UCP_EP_FLAG_REMOTE_ID) && /* remote ID defined */
-           (ep->flags & UCP_EP_FLAG_REMOTE_CONNECTED) && /* wireup completed */
-           (ucp_ep_get_iface_attr(ep, lane)->cap.flags &
+    return (ucp_ep->flags & UCP_EP_FLAG_REMOTE_ID) && /* remote ID defined */
+           (rsc_idx != UCP_NULL_RESOURCE) && /* rsc index is not NULL, i.e.
+                                                it is not a CM lane */
+           (ucp_worker_iface(ucp_ep->worker, rsc_idx)->attr.cap.flags &
             UCT_IFACE_FLAG_CONNECT_TO_IFACE); /* connect-to-iface */
 }
 
+ucs_status_t ucp_ep_do_uct_ep_keepalive(ucp_ep_h ucp_ep, uct_ep_h uct_ep,
+                                        ucp_rsc_index_t rsc_idx, unsigned flags,
+                                        uct_completion_t *comp)
+{
+    ucs_status_t status;
+    ucp_request_t req;
+    ssize_t packed_len;
+
+    ucs_assert((rsc_idx == UCP_NULL_RESOURCE) ||
+               (ucp_worker_iface(ucp_ep->worker, rsc_idx)->attr.cap.flags &
+                UCT_IFACE_FLAG_EP_CHECK));
+
+    if (ucp_ep_is_am_keepalive(ucp_ep, rsc_idx)) {
+        ucs_assert(ucp_worker_iface(ucp_ep->worker, rsc_idx)->attr.cap.flags &
+                   UCT_IFACE_FLAG_AM_BCOPY);
+        status = ucp_wireup_msg_prepare(&req, ucp_ep, UCP_WIREUP_MSG_EP_CHECK,
```

Comment:
why need to have a request object?
maybe change the context of ucp_wireup_msg_pack function that instead of request it will be a pointer to "struct iovec" array. the function will assume it has 2 elements (wireup msg and address).
need to update wireup_msg_progress to create such iovec[2] as local var and pass it to uct_am_bcopy
if needed the 2nd iovec len can be 0

Reply:
good idea
thank you
done